java-squid/src/main/java/org/sonar/java/se/ConstraintManager.java

```diff
@@ -0,0 +1,253 @@
+    SymbolicValue result = new SymbolicValue.ObjectSymbolicValue(map.size() + ProgramState.EMPTY_STATE.constraints.size());
+    map.put(variable, result);
+    symbolMap.put(variable.symbol(), result);
+    return result;
+  }
+
+  public SymbolicValue eval(ProgramState programState, Tree node) {
+    Tree syntaxNode = skipTrivial(node);
+    switch (syntaxNode.kind()) {
+      case NULL_LITERAL:
+        return SymbolicValue.NULL_LITERAL;
+      case BOOLEAN_LITERAL:
+        boolean value = Boolean.parseBoolean(((LiteralTree) syntaxNode).value());
+        if (value) {
+          return SymbolicValue.TRUE_LITERAL;
+        }
+        return SymbolicValue.FALSE_LITERAL;
+      case VARIABLE:
+        SymbolicValue variableResult = programState.values.get(((VariableTree) syntaxNode).symbol());
+        if (variableResult != null) {
+          // symbolic value associated with local variable
+          return variableResult;
+        }
+        break;
+      case IDENTIFIER:
+        SymbolicValue identifierResult = programState.values.get(((IdentifierTree) syntaxNode).symbol());
+        if (identifierResult != null) {
+          // symbolic value associated with local variable
+          return identifierResult;
+        }
+        break;
+      case LOGICAL_COMPLEMENT:
+        UnaryExpressionTree unaryExpressionTree = (UnaryExpressionTree) syntaxNode;
+        SymbolicValue val = eval(programState, unaryExpressionTree.expression());
+        if (SymbolicValue.FALSE_LITERAL.equals(val)) {
+          return SymbolicValue.TRUE_LITERAL;
+        } else if (val.equals(SymbolicValue.TRUE_LITERAL)) {
+          return SymbolicValue.FALSE_LITERAL;
+        }
+        // if not tied to a concrete value, create symbolic value with no constraint for now.
+        // TODO : create constraint between expression and created symbolic value
+      default:
+    }
+    return createSymbolicValue(syntaxNode);
+  }
+
+  /**
+   * Remove parenthesis and type cast.
+   */
+  private static Tree skipTrivial(Tree givenExpression) {
+    Tree expression = givenExpression;
+    do {
+      switch (expression.kind()) {
+        case PARENTHESIZED_EXPRESSION:
+          expression = ((ParenthesizedTree) expression).expression();
+          break;
+        case TYPE_CAST:
+          expression = ((TypeCastTree) expression).expression();
+          break;
+        default:
+          return expression;
+      }
+    } while (true);
+  }
+
+  public boolean isNull(ProgramState ps, SymbolicValue val) {
+    return NullConstraint.NULL.equals(ps.constraints.get(val));
+  }
+
+  public boolean isConstrained(ProgramState ps, SymbolicValue val) {
+    return ps.constraints.containsKey(val);
+  }
+
+  public Pair<ProgramState, ProgramState> assumeDual(ProgramState programState, Tree givenCondition) {
+    // FIXME condition value should be evaluated to determine if it is worth exploring this branch. This should probably be done in a
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Take the required action to fix the issue indicated by this comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1134)